Retaining unused imports based on user's choice

abhiashu opened the pull request:
Hi, We are trying to use autoimport in one of the huge project's python migration. I observed that default behaviour of autoimport is to fix missing import and also remove unused imports. 
The unused imports removal can cause problems sometimes because many times these imports are not directly accessed but dynamically accessed by creating and evaluating method/class names at runtime, hence these imports may appear as unused but are being actually used.
Similarly there may be cases where people import all methods/classes from a module using * and use only 1-2. Thus this * import may seem unused but actually being used.
I am suggesting an addition of a choice flag to retain unused imports based on user's choice. The default behaviour will not change i.e if user does not specify this flag autoimport will remove unused imports as it does today.
Have defined unit test as well but wasn't able to test because of below error-
`(python39_env) chaturvedi@Abhishekns-MacBook-Pro unit % python -m pytest test_services.py
ERROR: usage: __main__.py [options] [file_or_dir] [file_or_dir] [...]
__main__.py: error: unrecognized arguments: -n
  inifile: /Users/chaturvedi/Documents/Repos/autoimport/pyproject.toml
  rootdir: /Users/chaturvedi/Documents/Repos/autoimport

`

Please review and let me know if it is missing out anything. 

Thanks,
Abhishek Chaturvedi
<!--
Thank you for sending a pull request!

Please describe what the change is, trying to link it with open issues.
-->

## Checklist

* [x] Add test cases to all the changes you introduce
* [ ] Update the documentation for the changes

abhiashu commented:
UNIT TEST RESULTS-


`
(python39_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % grep "test_fix_remove_unused_imports" /tmp/test_results.log
tests/unit/test_services.py::test_fix_remove_unused_imports 
[gw0] [ 90%] PASSED tests/unit/test_services.py::test_fix_remove_unused_imports 
(python39_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % 
(python39_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % grep "test_fix_not_remove_unused_imports" /tmp/test_results.log
tests/unit/test_services.py::test_fix_not_remove_unused_imports 
[gw1] [100%] PASSED tests/unit/test_services.py::test_fix_not_remove_unused_imports 


---------- coverage: platform darwin, python 3.9.19-final-0 ----------
Name                                     Stmts   Miss  Cover   Missing
----------------------------------------------------------------------
src/autoimport/__init__.py                   3      0   100%
src/autoimport/__main__.py                   1      1     0%   3
src/autoimport/entrypoints/__init__.py      11      7    36%   20-29
src/autoimport/entrypoints/cli.py           54      3    94%   90, 96-97
src/autoimport/model.py                    234      1    99%   372
src/autoimport/services.py                  24      2    92%   44-45
src/autoimport/version.py                    6      0   100%
----------------------------------------------------------------------
TOTAL                                      333     14    96%


======================== 104 passed, 1 skipped in 2.40s ========================

`

lyz-code commented on tests/unit/test_services.py:
```diff
@@ -964,6 +964,71 @@ def test_fix_doesnt_fail_on_empty_file() -> None:
     assert result == source
 
 
+def test_fix_remove_unused_imports() -> None:
```
If this is the default behaviour then it's already covered in the rest of the tests. Thus I'll remove this one

lyz-code reviewed (changes requested):
Hi @abhiashu thank you for the contribution. I have mixed feelings about it though, as I don't  clearly understand the use cases behind the need not to remove unused imports.

> many times these imports are not directly accessed but dynamically accessed by creating and evaluating method/class names at runtime

Can you please share with me a snippet that represents the case above of dynamically creating and evaluating at runtime? I have never encountered such case and I'd like to understand it better.

> there may be cases where people import all methods/classes from a module using * and use only 1-2

This is a known antipattern that should be avoided at all costs for two reasons:

- If two libraries have the same class/function/object and you `import *` on both there may be conflicts
- It's difficult to know where does a class/function/object come from

Maybe there is a benefit  of using `import *` that I'm missing, if so let me know :). Being `autoimport` an opinionated tool, I'd rather not contribute in the spreading of this antipattern without a good reason.

There are other tools out there that does what you are trying to do ([`smart_imports`](https://github.com/Tiendil/smart-imports) is one that pops into my  mind), I also think the `isort` repository has an issue that lists some of them (you'll find it by searching `autoimport` in their repo issues). The differentiate factor of `autoimport` is that it does remove the unneeded ones. What I mean by this is that maybe you're better off using other tools that may fulfill your use case better.

Having said that as far as the changes made into the code are minimal and it doesn't add too much complexity I'm fine to accept the PR if you also update the docs to let the users know of the existence of the flag and make the tests pass.

Remember that you can use `make all` and `make test` locally to run the checks that the CI runs.

Thanks again :)

abhiashu commented:
Hi @lyz-code I have made the changes as you suggested, also made changes suggested by make all command. But I am facing issues in committing files. I tried with both python3.9 and 3.7 virtual environments-
```
(python37_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % git commit -m "Removed unwanted test case, make all changes addressed."  
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /Users/chaturvedi/.cache/pre-commit/patch1715140750-55299.
Trim Trailing Whitespace.................................................Passed
Check for added large files..............................................Passed
Check docstring is first.................................................Passed
Check for merge conflicts................................................Passed
Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing .idea/misc.xml

black....................................................................Passed
Run mypy static analysis tool............................................Failed
- hook id: mypy
- exit code: 2

pyproject.toml:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'  [misc]
Found 1 error in 1 file (errors prevented further checking)

[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
[INFO] Restored changes from /Users/chaturvedi/.cache/pre-commit/patch1715140750-55299.


(python37_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % pip freeze | grep pydantic
pydantic==2.5.3
pydantic_core==2.14.6
(python37_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % pip freeze | grep mypy    
mypy==1.4.1
mypy-extensions==1.0.0
```

Python3.9

```
(python39_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % git commit -m "Removed unwanted test case, make all changes addressed."  
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /Users/chaturvedi/.cache/pre-commit/patch1715140932-55446.
Trim Trailing Whitespace.................................................Passed
Check for added large files..............................................Passed
Check docstring is first.................................................Passed
Check for merge conflicts................................................Passed
Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing .idea/misc.xml

black....................................................................Passed
Run mypy static analysis tool............................................Failed
- hook id: mypy
- exit code: 2

pyproject.toml:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'  [misc]
Found 1 error in 1 file (errors prevented further checking)

[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
[INFO] Restored changes from /Users/chaturvedi/.cache/pre-commit/patch1715140932-55446.


(python39_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % pip freeze | grep pydantic                                               
pydantic==1.10.13
pydantic-factories==1.17.3
pydantic_core==2.18.2
(python39_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % pip freeze | grep mypy                                                   
mypy==1.7.0
mypy-extensions==1.0.0
```

pydantic and mypy are installed in both virtual environments, please clarify this issue.

lyz-code commented:
Hi @abhiashu thanks for your clarifying answer. I understand now the case of the dynamic imports and I agree with you that supporting this case will be nice. Your proposal to use the flag looks sound to me too.

> Also I have noted another point, that autoimport does not fix imports spread over multiple lines

In theory it should, it's been a while since I last delved into the code so I can't guarantee it, but if it doesn't and you want to contribute a fix, it will be more than welcome :)

> I tried with both python3.9 and 3.7 virtual environments ...

Python 3.7 is deprecated so I wouldn't use time trying to debug the issue in that version.

> pyproject.toml:1: error: Error importing plugin "pydantic.mypy": No module named 'pydantic'  [misc]
Found 1 error in 1 file (errors prevented further checking)

It looks like the `pre-commit` mypy hook is failing, I'm not sure why as I really don't use the pre-commits (I should remove them I guess, but I thought it may be useful to some users). You can get rid of the issue running `pre-commit --uninstall`. The test that is important (the one run by the CI) is `make mypy`, if that passes then it's all good.

Best,
Lyz

abhiashu commented:
Hi @lyz-code thanks for your reply.
I have updated the documentation and pushed the changes as per make all lint recommendations as well. Also, removed the unwanted test cases as you had mentioned. Please have a look.

Test case run-
```
---------- coverage: platform darwin, python 3.9.19-final-0 ----------
Name                                     Stmts   Miss  Cover   Missing
----------------------------------------------------------------------
src/autoimport/__init__.py                   3      3     0%   8-12
src/autoimport/__main__.py                   1      1     0%   3
src/autoimport/entrypoints/__init__.py      11     11     0%   7-29
src/autoimport/entrypoints/cli.py           54     54     0%   3-105
src/autoimport/model.py                    234    234     0%   3-561
src/autoimport/services.py                  24     24     0%   7-80
src/autoimport/version.py                    6      6     0%   3-14
----------------------------------------------------------------------
TOTAL                                      333    333     0%


=============================================================================================== 103 passed, 1 skipped in 2.56s ===============================================================================================

--------------------
- Testing examples -
--------------------
# @find docs/examples -type f -name '*.py' | xargs -I'{}' sh -c 'echo {}; pdm run python {} >/dev/null 2>&1 || (echo "{} failed" ; exit 1)'

# pdm run pytest docs/examples/*
```

mypy test-
```
(python39_env) chaturvedi@Abhishekns-MacBook-Pro autoimport % make mypy
----------------
- Testing mypy -
----------------
pdm run mypy src tests
INFO: Inside an active virtualenv /Users/chaturvedi/Documents/virtual_envs/python39_env, reusing it.
Set env var PDM_IGNORE_ACTIVE_VENV to ignore it.
Success: no issues found in 16 source files
```

Please review and let me know.

lyz-code commented on docs/index.md:
```diff
@@ -128,6 +128,12 @@ This can be problematic when run in `__init__.py` files, which often contain
+There may be import statements that are being used dynamically, 
+to autoimport it would look like those are not being used but actually they may have some usages.
+
+In such cases where you want to retain the unused imports across any file, you can use the `--keep-unused-imports` flag, 
```
I'd add a note that if there are not that many cases, using  `# noqa: autoimport` should be favoured over using the flag.

lyz-code reviewed (changes requested):
Hi @abhiashu one little last tweak and we're good to go :)

lyz-code commented:
Damn it looks like the CI is failing because some of the dependencies have vulnerabilities. This is because since I entered in low maintenance mode on this repo the [update pipeline](https://github.com/lyz-code/autoimport/actions) has not been maintained.

I know it doesn't have anything to do with your PR, but I don't have time to fix it myself.

So if you have the energies to work on it I suggest you run `make update-production` to update the minimum dependencies and see if it gets fixed. If it doesn't then you can run `make update`.

To test that the CI will pass you can run `make security`. Remember that to run all tests you can run `make all`

abhiashu commented:
> Damn it looks like the CI is failing because some of the dependencies have vulnerabilities. This is because since I entered in low maintenance mode on this repo the [update pipeline](https://github.com/lyz-code/autoimport/actions) has not been maintained.
> 
> I know it doesn't have anything to do with your PR, but I don't have time to fix it myself.
> 
> So if you have the energies to work on it I suggest you run `make update-production` to update the minimum dependencies and see if it gets fixed. If it doesn't then you can run `make update`.
> 
> To test that the CI will pass you can run `make security`. Remember that to run all tests you can run `make all`

Hi @lyz-code 
I have updated the documentation as requested.

Updated the dependencies as well, please check. Post that also the pipeline was failing. Upon debugging I found that my test case has csv23 mentioned, which may not be installed on the pipeline server because of which it was not able to add it in the test code hence that test case was failing. After modifying csv23 to csv, it worked. From this I have another doubt - to add any missing import does that package need to be installed beforehand otherwise autoimport won't add its import?

All checks have passed now, please review the changes.

lyz-code commented:
Thanks for fixing everything!

> to add any missing import does that package need to be installed beforehand otherwise autoimport won't add its import?

Yes it's needed for how autoimport works

lyz-code commented:
Released under `1.5.0` thanks for all the work :)

abhiashu commented:
Thanks a lot @lyz-code
